lib/rucio/client/downloadclient.py

```diff
@@ -73,6 +73,134 @@ def __init__(self, _client=None, user_agent='rucio_clients', logger=None):
+                pfn_archive = pfns_dict[archive]
+            except:
+                raise InputValidationError('The pfn of archive file not found.')
+
+            if '*' in archive:
+                logger.debug(archive)
+                raise InputValidationError('Cannot use PFN download with wildcard in DID')
+
+            # preparing output directories
+            dest_dir_path = self._prepare_dest_dir(item.get('base_dir', '.'),
+                                                   os.path.join(archive_scope, archive_name + '.extracted'), file_extract,
+                                                   item.get('no_subdir'))
+            logger.debug('Preparing output destination %s' % dest_dir_path)
+
+            item['file_extract'] = file_extract_name
+            item['scope_archive'] = archive_scope
+            item['name_archive'] = archive_name
+            item['rse'] = rse_name
+            item['pfn_archive'] = pfn_archive
+            item['force_scheme'] = pfn_archive.split(':')[0]
+            item['dest_dir_path'] = dest_dir_path
+            item.setdefault('ignore_checksum', True)
+            input_items.append(item)
+
+        for item in input_items:
+            trace['uuid'] = generate_uuid()
+            trace['scope'] = item['scope_archive']
+            trace['dataset'] = item['name_archive']
+            trace['filename'] = item['file_extract']
+            trace['rse'] = item['rse']
+
+            # if file already exists, set state, send trace, and return
+            dest_dir_path = item['dest_dir_path']
+            dest_file_path = os.path.join(dest_dir_path, file_extract)
+            if os.path.isfile(dest_file_path):
+                logger.info('%s File exists already locally: %s' % (file_extract, dest_dir_path))
+                trace['clientState'] = 'ALREADY_DONE'
+                trace['transferStart'] = time.time()
+                trace['transferEnd'] = time.time()
+                send_trace(trace, self.client.host, self.user_agent)
+                continue
+            try:
+                start_time = time.time()
+                cmd = 'xrdcp -vf %s -z %s file://%s' % (item['pfn_archive'], item['file_extract'], item['dest_dir_path'])
+                status, out, err = execute(cmd)
+                end_time = time.time()
+                trace['transferStart'] = start_time
+                trace['transferEnd'] = end_time
+                if status == 54:
+                    trace['clientState'] = 'FAILED'
+                    raise SourceNotFound()
+                elif status != 0:
+                    trace['clientState'] = 'FAILED'
+                    raise RucioException(err)
+                else:
+                    trace['clientState'] = 'DONE'
+            except Exception as e:
+                trace['clientState'] = 'FAILED'
+                raise ServiceUnavailable(e)
+            send_trace(trace, self.client.host, self.user_agent)
+
```

Comment:
No checksum validation ?

Reply:
we have no metadata for files inside archive yet, I'll add checksum validation as soon as we have metadata.